File: scripts/fix-dollar-one.js
#!/usr/bin/env node
/**
 * Fix $1 placeholders left by bad regex replacement
 *
 * This script systematically fixes broken code patterns where "$1" appears
 * instead of properly expanded capture groups.
 */

const fs = require("node:fs");
const path = require("node:path");
const { glob } = require("glob");

// Pattern fixes - ordered by specificity (most specific first)
const fixes = [
	// Database operations
	{ pattern: /\.selec"\$1"/g, replacement: ".select()" },
	{ pattern: /\.inser"\$1"/g, replacement: ".insert" }, // Note: needs table name after
	{ pattern: /\.se"\$1"/g, replacement: ".set({" }, // Partial - context dependent

	// Test patterns
	{
		pattern: /\si"\$1"\s*=>\s*\{/g,
		replacement: ' it("test case", async () => {',
	},
	{
		pattern: /expec"\$1"\.toBe\(/g,
		replacement: "expect(response.status).toBe(",
	},
	{
		pattern: /expec"\$1"\.toMatchObjec"\$1"/g,
		replacement: "expect(data).toMatchObject({",
	},
	{
		pattern: /expec"\$1"\.toBeDefined\(\)/g,
		replacement: "expect(data).toBeDefined()",
	},
	{
		pattern: /expec"\$1"\.toBeNull\(\)/g,
		replacement: "expect(trial.userId).toBeNull()",
	},
	{
		pattern: /expec"\$1"\.toHaveLength\(/g,
		replacement: "expect(trials).toHaveLength(",
	},
	{
		pattern: /expec"\$1"\.toHaveProperty\(/g,
		replacement: "expect(data).toHaveProperty(",
	},
	{
		pattern: /expec"\$1"\.toContain\(/g,
		replacement: "expect(data.errors).toContain(",
	},
	{
		pattern: /expec"\$1"\.toMatch\(/g,
		replacement: "expect(data.apiKey).toMatch(",
	},
	{
		pattern: /expec"\$1"\)\.toBeGreaterThan\(/g,
		replacement: "expect(new Date(data.blockedUntil)).toBeGreaterThan(",
	},
	{
		pattern: /expec"\$1"\.getTime\(\)\)\.toBeGreaterThan\(/g,
		replacement: "expect(new Date(data.blockedUntil).getTime()).toBeGreaterThan(",
	},
	{
		pattern: /findFirs"\$1"/g,
		replacement: "findFirst({ where: eq(deviceTrials.deviceFingerprint, deviceFingerprint) })",
	},

	// Function signatures
	{
		pattern: /async function (\w+)"\$1"/g,
		replacement: "async function $1()",
	},
	{
		pattern: /export async function (\w+)"\$1"/g,
		replacement: "export async function $1()",
	},
	{ pattern: /async (\w+)"\$1":/g, replacement: "async $1(" },
	{ pattern: /trackDeviceEven"\$1"/g, replacement: "trackDeviceEvent" },
	{ pattern: /trackUserEven"\$1"/g, replacement: "trackUserEvent" },
	{ pattern: /captureEven"\$1"/g, replacement: "capture" },
	{
		pattern: /getClientFingerprin"\$1"/g,
		replacement: "getClientFingerprint()",
	},
	{ pattern: /createClien"\$1"/g, replacement: "createClient()" },
	{
		pattern: /updateCheckpointLimi"\$1"/g,
		replacement: "updateCheckpointLimit",
	},
	{ pattern: /trackPaymen"\$1"/g, replacement: "trackPayment" },
	{ pattern: /sendPaymentReceip"\$1"/g, replacement: "sendPaymentReceipt" },
	{ pattern: /suspendAccoun"\$1"/g, replacement: "suspendAccount" },

	// Web API calls
	{ pattern: /\.charCodeA"\$1"/g, replacement: ".charCodeAt(i)" },
	{ pattern: /\.padStar"\$1"/g, replacement: ".padStart(2, '0')" },
	{ pattern: /\.diges"\$1"/g, replacement: ".digest('SHA-256', dataBuffer)" },
	{
		pattern: /Intl\.DateTimeForma"\$1"\.resolvedOptions\(\)\.timeZone/g,
		replacement: "Intl.DateTimeFormat().resolvedOptions().timeZone",
	},
	{
		pattern: /request\.headers\.ge"\$1"/g,
		replacement: "request.headers.get",
	},
	{ pattern: /request\.tex"\$1"/g, replacement: "request.text()" },
	{ pattern: /rese"\$1"/g, replacement: "reset()" },

	// Currency (special case - this is a stub)
	{
		pattern: /return "\$1,000\.00";/g,
		replacement: 'return "$1,000.00"; // Stub implementation',
	},
];

async function fixFile(filePath) {
	let content = fs.readFileSync(filePath, "utf8");
	let modified = false;

	for (const fix of fixes) {
		const before = content;
		content = content.replace(fix.pattern, fix.replacement);
		if (content !== before) {
			modified = true;
		}
	}

	if (modified) {
		fs.writeFileSync(filePath, content, "utf8");
		return true;
	}

	return false;
}

async function main() {
	// Parse CLI arguments
	const args = process.argv.slice(2);
	const patternsArg = args.find((arg) => arg.startsWith("--patterns="));
	const patterns = patternsArg
		? patternsArg.split("=")[1].split(",")
		: ["apps/web/**/*.ts", "apps/web/**/*.tsx", "!apps/web/node_modules/**", "!apps/web/.next/**"];

	const files = await glob(patterns, { absolute: true });
	const filesToFix = [];

	// First pass: identify files with $1 issues
	for (const file of files) {
		const content = fs.readFileSync(file, "utf8");
		if (content.includes("$1")) {
			filesToFix.push(file);
		}
	}

	console.log(`Found ${filesToFix.length} files with $1 issues`);

	let fixedCount = 0;
	for (const file of filesToFix) {
		const wasFixed = await fixFile(file);
		if (wasFixed) {
			fixedCount++;
			console.log(`✓ Fixed: ${path.relative(process.cwd(), file)}`);
		}
	}

	console.log(`\nFixed ${fixedCount} files`);

	// Check if any $1 instances remain
	const remaining = [];
	for (const file of files) {
		const content = fs.readFileSync(file, "utf8");
		if (content.includes("$1")) {
			const lines = content.split("\n");
			lines.forEach((line, idx) => {
				if (line.includes("$1")) {
					remaining.push({
						file,
						line: idx + 1,
						content: line.trim(),
					});
				}
			});
		}
	}

	if (remaining.length > 0) {
		console.log(`\n⚠️  ${remaining.length} instances of $1 still remain:`);
		remaining.slice(0, 10).forEach((r) => {
			console.log(`  ${path.relative(process.cwd(), r.file)}:${r.line} - ${r.content.substring(0, 80)}`);
		});
		if (remaining.length > 10) {
			console.log(`  ... and ${remaining.length - 10} more`);
		}
	} else {
		console.log("\n✅ All $1 instances have been fixed!");
	}

	// Show help if requested
	if (args.includes("--help") || args.includes("-h")) {
		console.log(`
Usage: node fix-dollar-one.js [options]

Options:
  --patterns=<glob>    Comma-separated glob patterns to search (default: apps/web/**/*.ts,apps/web/**/*.tsx,!apps/web/node_modules/**,!apps/web/.next/**)
  --help, -h          Show this help message

Examples:
  node fix-dollar-one.js
  node fix-dollar-one.js --patterns="packages/**/*.ts,!packages/node_modules/**"
		`);
	}
}

main().catch(console.error);

File: scripts/fix-dollar-one.py
#!/usr/bin/env python3
"""
Fix $1 placeholders left by bad regex replacement

This script systematically fixes broken code patterns where "$1" appears
instead of properly expanded capture groups.
"""

import os
import re
from pathlib import Path

# Define fix patterns - order matters (most specific first)
FIXES = [
    # Function/method patterns
    (r'export default function (\w+)"\$1"', r'export default function \1()'),
    (r'export async function (\w+)"\$1"', r'export async function \1()'),
    (r'export function (\w+)"\$1"', r'export function \1()'),
    (r'async function (\w+)"\$1"', r'async function \1()'),
    (r'function (\w+)"\$1"', r'function \1()'),
    (r'(\w+)"\$1"\s*:\s*Promise<', r'\1('),  # async method signatures

    # Test patterns
    (r'\bit"\$1"\s*=>\s*\{', r'it("test", async () => {'),
    (r'\bdescribe"\$1"\s*=>\s*\{', r'describe("test", () => {'),

    # Database operations (Drizzle ORM)
    (r'\.select"\$1"', r'.select()'),
    (r'\.insert"\$1"', r'.insert'),
    (r'\.update"\$1"', r'.update'),
    (r'\.delete"\$1"', r'.delete'),
    (r'\.set"\$1"', r'.set({'),
    (r'\.where"\$1"', r'.where('),
    (r'\.findFirst"\$1"', r'.findFirst({'),
    (r'\.findMany"\$1"', r'.findMany({'),

    # expect() patterns - context dependent, use generic data
    (r'expect"\$1"\.toBe\(', r'expect(data).toBe('),
    (r'expect"\$1"\.toEqual\(', r'expect(data).toEqual('),
    (r'expect"\$1"\.toMatchObject"\$1"', r'expect(data).toMatchObject({'),
    (r'expect"\$1"\.toBeDefined\(\)', r'expect(data).toBeDefined()'),
    (r'expect"\$1"\.toBeNull\(\)', r'expect(data).toBeNull()'),
    (r'expect"\$1"\.toBeTruthy\(\)', r'expect(data).toBeTruthy()'),
    (r'expect"\$1"\.toBeFalsy\(\)', r'expect(data).toBeFalsy()'),
    (r'expect"\$1"\.toBeUndefined\(\)', r'expect(data).toBeUndefined()'),
    (r'expect"\$1"\.toHaveLength\(', r'expect(data).toHaveLength('),
    (r'expect"\$1"\.toHaveProperty\(', r'expect(data).toHaveProperty('),
    (r'expect"\$1"\.toContain\(', r'expect(data).toContain('),
    (r'expect"\$1"\.toMatch\(', r'expect(data).toMatch('),
    (r'expect"\$1"\.not\.toHaveProperty\(', r'expect(data).not.toHaveProperty('),
    (r'expect"\$1"\)\.toBeGreaterThan\(', r'expect(data).toBeGreaterThan('),
    (r'expect"\$1"\.getTime\(\)\)\.toBeGreaterThan\(', r'expect(new Date(data).getTime()).toBeGreaterThan('),

    # Specific method names
    (r'trackDeviceEvent"\$1"', r'trackDeviceEvent('),
    (r'trackUserEvent"\$1"', r'trackUserEvent('),
    (r'captureEvent"\$1"', r'capture({'),
    (r'getClientFingerprint"\$1"', r'getClientFingerprint()'),
    (r'createClient"\$1"', r'createClient()'),
    (r'updateCheckpointLimit"\$1"', r'updateCheckpointLimit('),
    (r'trackPayment"\$1"', r'trackPayment('),
    (r'sendPaymentReceipt"\$1"', r'sendPaymentReceipt('),
    (r'suspendAccount"\$1"', r'suspendAccount('),

    # Web API patterns
    (r'\.charCodeAt"\$1"', r'.charCodeAt(i)'),
    (r'\.padStart"\$1"', r'.padStart(2, \'0\')'),
    (r'\.digest"\$1"', r'.digest(\'SHA-256\', dataBuffer)'),
    (r'Intl\.DateTimeFormat"\$1"\.resolvedOptions\(\)\.timeZone', r'Intl.DateTimeFormat().resolvedOptions().timeZone'),
    (r'request\.headers\.get"\$1"', r'request.headers.get('),
    (r'request\.text"\$1"', r'request.text()'),
    (r'request\.json"\$1"', r'request.json()'),
    (r'response\.json"\$1"', r'response.json()'),
    (r'reset"\$1"', r'reset()'),

    # useEffect and React hooks
    (r'useEffect"\$1"', r'useEffect(() => {'),
    (r'useState"\$1"', r'useState('),
    (r'useCallback"\$1"', r'useCallback(() => {'),
    (r'useMemo"\$1"', r'useMemo(() => {'),
    (r'useRef"\$1"', r'useRef('),

    # Common method calls
    (r'\.map"\$1"', r'.map('),
    (r'\.filter"\$1"', r'.filter('),
    (r'\.reduce"\$1"', r'.reduce('),
    (r'\.forEach"\$1"', r'.forEach('),
    (r'\.find"\$1"', r'.find('),
    (r'\.some"\$1"', r'.some('),
    (r'\.every"\$1"', r'.every('),

    # Type assertions and generics
    (r'<\w+>"\$1"', r'<T>()'),
]

def fix_file(file_path):
    """Fix $1 instances in a single file"""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()

        # Skip if no $1 found
        if '$1' not in content:
            return False

        original = content

        # Apply all fixes
        for pattern, replacement in FIXES:
            content = re.sub(pattern, replacement, content)

        # Write back if changed
        if content != original:
            with open(file_path, 'w', encoding='utf-8') as f:
                f.write(content)
            return True

        return False
    except Exception as e:
        print(f"Error processing {file_path}: {e}")
        return False

def main():
    """Main execution"""
    base_dir = Path(__file__).parent.parent / 'apps' / 'web'

    # Find all TS/TSX files
    files = list(base_dir.rglob('*.ts')) + list(base_dir.rglob('*.tsx'))

    # Filter out node_modules and .next
    files = [f for f in files if 'node_modules' not in str(f) and '.next' not in str(f)]

    # Find files with $1
    affected_files = []
    for file in files:
        try:
            with open(file, 'r', encoding='utf-8') as f:
                if '$1' in f.read():
                    affected_files.append(file)
        except:
            pass

    print(f"Found {len(affected_files)} files with $1 issues")

    # Fix each file
    fixed_count = 0
    for file in affected_files:
        if fix_file(file):
            fixed_count += 1
            rel_path = file.relative_to(base_dir.parent.parent)
            print(f"✓ Fixed: {rel_path}")

    print(f"\nFixed {fixed_count} files")

    # Check remaining issues
    remaining = []
    for file in files:
        try:
            with open(file, 'r', encoding='utf-8') as f:
                lines = f.readlines()
                for i, line in enumerate(lines, 1):
                    if '$1' in line:
                        remaining.append((file, i, line.strip()))
        except:
            pass

    if remaining:
        print(f"\n⚠️  {len(remaining)} instances of $1 still remain")
        for file, line_no, line in remaining[:20]:
            rel_path = file.relative_to(base_dir.parent.parent)
            print(f"  {rel_path}:{line_no} - {line[:80]}")
        if len(remaining) > 20:
            print(f"  ... and {len(remaining) - 20} more")
    else:
        print("\n✅ All $1 instances have been fixed!")

if __name__ == '__main__':
    main()
